##### Homework
> Note: sometimes your answer doesn't match one of the options exactly. 
> That's fine. 
> Select the option that's closest to your solution.
> Note: we recommend using python 3.11 in this homework.
##### Question 1

* Install Pipenv
* What's the version of pipenv you installed?
* Use `--version` to find out


##### Question 2

* Use Pipenv to install Scikit-Learn version 1.5.2
* What's the first hash for scikit-learn you get in Pipfile.lock?
> **Note**: you should create an empty folder for homework  and do it there. 


##### Models

We've prepared a dictionary vectorizer and a model.

They were trained (roughly) using this code:

```python
features = ['job', 'duration', 'poutcome']
dicts = df[features].to_dict(orient='records')

dv = DictVectorizer(sparse=False)
X = dv.fit_transform(dicts)

model = LogisticRegression().fit(X, y)
```

> **Note**: You don't need to train the model. This code is just for your reference.

And then saved with Pickle. Download them:

* [DictVectorizer](https://github.com/DataTalksClub/machine-learning-zoomcamp/tree/master/cohorts/2024/05-deployment/homework/dv.bin?raw=true)
* [LogisticRegression](https://github.com/DataTalksClub/machine-learning-zoomcamp/tree/master/cohorts/2024/05-deployment/homework/model1.bin?raw=true)

With `wget`:

```bash
PREFIX=https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2024/05-deployment/homework
wget $PREFIX/model1.bin
wget $PREFIX/dv.bin
```


In [17]:
import os
os.getcwd().split('\\')[-2:]  


['cohorts', '05-deployment']

In [14]:
!mkdir homework

PREFIX="https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2024/05-deployment/homework"
!wget -P homework $PREFIX/model1.bin 

!wget -P homework $PREFIX/dv.bin
# ```

�l�ؿ����ɮ� homework �w�g�s�b�C
--2024-10-29 00:17:27--  https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2024/05-deployment/homework/model1.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 850 [application/octet-stream]
Saving to: 'homework/model1.bin'

     0K                                                       100% 18.6M=0s

2024-10-29 00:17:28 (18.6 MB/s) - 'homework/model1.bin' saved [850/850]

--2024-10-29 00:17:28--  https://raw.githubusercontent.com/DataTalksClub/machine-learning-zoomcamp/master/cohorts/2024/05-deployment/homework/dv.bin
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githu

In [18]:
os.chdir('./homework')
os.getcwd().split('\\')[-2:]  


['05-deployment', 'homework']

In [20]:
# Question 1
!C:\Python311\python.exe -m pipenv --version

pipenv, version 2024.0.1


In [21]:
# Question 2
!C:\Python311\python.exe -m pipenv install scikit-learn==1.5.2


[    ] Creating virtual environment...
[=   ] Creating virtual environment...
[==  ] Creating virtual environment...
[=== ] Creating virtual environment...
[ ===] Creating virtual environment...
[  ==] Creating virtual environment...
[    ] Creating virtual environment...
[   =] Creating virtual environment...
[  ==] Creating virtual environment...
[====] Creating virtual environment...
[=== ] Creating virtual environment...
[==  ] Creating virtual environment...
[    ] Creating virtual environment...
[=   ] Creating virtual environment...
[==  ] Creating virtual environment...
[=== ] Creating virtual environment...
[  ==] Creating virtual environment...
[   =] Creating virtual environment...
[    ] Creating virtual environment...
[  ==] Creating virtual environment...
[ ===] Creating virtual environment...
[====] Creating virtual environment...
[=== ] Creating virtual environment...
[==  ] Creating virtual environment...
[    ] Creating virtual environment...
[=   ] Creating virtual e

Creating a virtualenv for this project...
Pipfile: C:\Users\wufen\Downloads\11\machine-learning-zoomcamp\cohorts\05-deployment\homework\Pipfile
Using default python from C:\Python311\python.exe (3.11.2) to create virtualenv...
created virtual environment CPython3.11.2.final.0-64 in 17991ms
  creator CPython3Windows(dest=C:\Users\wufen\.virtualenvs\homework-laXm1tgw, clear=False, no_vcs_ignore=False, global=False)
  seeder FromAppData(download=False, pip=bundle, setuptools=bundle, wheel=bundle, via=copy, app_data_dir=C:\Users\wufen\AppData\Local\pypa\virtualenv)
    added seed packages: pip==24.2, setuptools==74.1.2, wheel==0.44.0
  activators BashActivator,BatchActivator,FishActivator,NushellActivator,PowerShellActivator,PythonActivator

Successfully created virtual environment!
Virtualenv location: C:\Users\wufen\.virtualenvs\homework-laXm1tgw
Creating a Pipfile for this project...
Pipfile.lock not found, creating...
Locking [packages] dependencies...
Locking [dev-packages] dependenci

In [23]:
import json

# 讀取 Pipfile.lock
with open('Pipfile.lock', 'r') as f:
    lockfile = json.load(f)

# 獲取 scikit-learn 的第一個 hash 值
scikit_learn_first_hash = lockfile['default']['scikit-learn']['hashes'][0]
print(f"scikit-learn 的第一個 hash 值是: {scikit_learn_first_hash}")

scikit-learn 的第一個 hash 值是: sha256:03b6158efa3faaf1feea3faa884c840ebd61b6484167c711548fce208ea09445


##### Question 3
Let's use these models!

* Write a script for loading these models with pickle
* Score this client:

```json
{"job": "management", "duration": 400, "poutcome": "success"}
```

What's the probability that this client will get a subscription? 

* 0.359
* 0.559
* 0.759
* 0.959


In [36]:
!C:\Python311\python.exe -m pipenv run python predict.py

0.7590966516879658



If you're getting errors when unpickling the files, check their checksum:

```bash

$ md5sum model1.bin dv.bin
3d8bb28974e55edefa000fe38fd3ed12  model1.bin
7d37616e00aa80f2152b8b0511fc2dff  dv.bin
```


In [28]:
import hashlib

def calculate_md5(filename):
    """計算文件的 MD5 校驗和"""
    md5_hash = hashlib.md5()
    with open(filename, "rb") as f:
        # 分塊讀取文件以處理大文件
        for chunk in iter(lambda: f.read(4096), b""):
            md5_hash.update(chunk)
    return md5_hash.hexdigest()


print("model1.bin MD5:", calculate_md5('model1.bin'))
print("dv.bin MD5:", calculate_md5('dv.bin'))

model1.bin MD5: 3d8bb28974e55edefa000fe38fd3ed12
dv.bin MD5: 7d37616e00aa80f2152b8b0511fc2dff


In [38]:
!C:\Python311\python.exe -m pipenv install flask waitress


Installing flask...
Resolving flask...
[    ] Installing...
Added flask to Pipfile's [packages] ...
[    ] Installing flask...
Installation Succeeded
[    ] Installing flask...
[    ] Installing flask...

Installing waitress...
Resolving waitress...
[    ] Installing...
Added waitress to Pipfile's [packages] ...
[    ] Installing waitress...
Installation Succeeded
[    ] Installing waitress...
[    ] Installing waitress...

Running $ pipenv lock then $ pipenv sync.
Building requirements...
[    ] Locking packages...
Resolving dependencies...
[    ] Locking packages...
[=   ] Locking packages...
[==  ] Locking packages...
[=== ] Locking packages...
[ ===] Locking packages...
[  ==] Locking packages...
[   =] Locking packages...
[   =] Locking packages...
[   =] Locking packages...
[ ===] Locking packages...
[====] Locking packages...
[=== ] Locking packages...
[==  ] Locking packages...
[    ] Locking packages...
[=   ] Locking packages...
[==  ] Locking packages...
[=== ] Locking packa

Pipfile.lock (344639) out of date: run `pipfile lock` to update to (1ee68c)...
Locking [packages] dependencies...
Locking [dev-packages] dependencies...
Updated Pipfile.lock (9d06ea120b959fb5be2febbfff8abb4c1c66908a97a24ff17ff6b621771ee68c)!


In [65]:
#!C:\Python311\python.exe -m pipenv run python predict_Q4.py
import requests

url = "http://localhost:9696/predict"
client = {"job": "student", "duration": 280, "poutcome": "failure"}
response = requests.post(url, json=client).json()
print("Subscription probability:", response['subscription_probability'])
print("subscribe:", response['subscribe'])



Subscription probability: 0.33480703475511053
subscribe: False


In [60]:
# response


In [66]:
# import requests

# url = "http://localhost:9696/predict"
# client = {"job": "management", "duration": 400, "poutcome": "success"}
# response = requests.post(url, json=client).json()
# print(response)

{'subscribe': True, 'subscription_probability': 0.7590966516879658}



##### Question 4

Now let's serve this model as a web service

* Install Flask and gunicorn (or waitress, if you're on Windows)
* Write Flask code for serving the model
* Now score this client using `requests`:

```python
url = "YOUR_URL"
client = {"job": "student", "duration": 280, "poutcome": "failure"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a subscription?

* 0.335
* 0.535
* 0.735
* 0.935


In [69]:
import requests
url = "http://localhost:9696/predict"
client = {"job": "student", "duration": 280, "poutcome": "failure"}
response = requests.post(url, json=client).json()
print(response)

{'subscribe': False, 'subscription_probability': 0.33480703475511053}


##### Docker
Install [Docker](https://github.com/DataTalksClub/machine-learning-zoomcamp/blob/master/05-deployment/06-docker.md). 
We will use it for the next two questions.

For these questions, we prepared a base image: `svizor/zoomcamp-model:3.11.5-slim`. 
You'll need to use it (see Question 5 for an example).

This image is based on `python:3.11.5-slim` and has a logistic regression model 
(a different one) as well a dictionary vectorizer inside. 

This is how the Dockerfile for this image looks like:

```docker 
FROM python:3.11.5-slim
WORKDIR /app
COPY ["model2.bin", "dv.bin", "./"]
```

We already built it and then pushed it to [`svizor/zoomcamp-model:3.11.5-slim`](https://hub.docker.com/r/svizor/zoomcamp-model).

> **Note**: You don't need to build this docker image, it's just for your reference.




##### Question 5

Download the base image `svizor/zoomcamp-model:3.11.5-slim`. You can easily make it by using [docker pull](https://docs.docker.com/engine/reference/commandline/pull/) command.

So what's the size of this base image?

* 45 MB
* 130 MB (0)
* 245 MB
* 330 MB

You can get this information when running `docker images` - it'll be in the "SIZE" column.




##### Dockerfile

Now create your own Dockerfile based on the image we prepared.

It should start like that:

```docker
FROM svizor/zoomcamp-model:3.11.5-slim
# add your stuff here
```

Now complete it:

* Install all the dependencies form the Pipenv file
* Copy your Flask script
* Run it with Gunicorn 

After that, you can build your docker image.


##### Question 6

Let's run your docker container!

After running it, score this client once again:

```python
url = "YOUR_URL"
client = {"job": "management", "duration": 400, "poutcome": "success"}
requests.post(url, json=client).json()
```

What's the probability that this client will get a subscription now?

* 0.287
* 0.530
* 0.757
* 0.960


In [73]:
import requests
url = "http://localhost:9696/predict"
client = {"job": "management", "duration": 400, "poutcome": "success"}
response = requests.post(url, json=client).json()
print(response)

{'subscribe': True, 'subscription_probability': 0.7590966516879658}


In [ ]:


# ## Submit the results

# * Submit your results here: https://courses.datatalks.club/ml-zoomcamp-2024/homework/hw05
# * If your answer doesn't match options exactly, select the closest one